docs/agents/log-shipper.md

```diff
@@ -0,0 +1,103 @@
+* Centralized logging for agent logs - better supportability
+* Drive awareness for APM in our logging user base
+  The easiest way to get started with centralized logging is to let the APM agent automagically do it
+
+## What about Elastic Agent?
+* Elastic Agent is not replacing standalone Java agent
+* Not having a dependency on Elastic Agent lets us do it now
+* When Elastic Agent comes along, we can also integrate with it
+* Filebeat can be used to ship the logs instead of the APM Agent for more advanced use cases or better delivery guarantees
+* If we see that shipping the logs from the Agent doesn't have value anymore because Elastic Agent is the standard for installing the Java agent, we can deprecate it.
+
+## Why log to a file vs send logs directly over the network?
+* When APM Server is down for a while, the log file acts as a buffer
+* The agent continues sending when the server gets back up again
+* If the application crashes logs are sent when it restarts
+* Lower memory footprint as logs don't have to be buffered in memory
+
+## Implementation details
+
+### Log tailing
+* The agent starts a background thread that constantly looks for new data in the file(s)
+* If there is new data, it writes it to the HTTP request body using a reusable byte buffer
+* The log tailing is designed to work with multiple files
+* Works with log rotation
+  * If the file has been fully read, it checks if there's a new one
+
+### Persistent state
+* Creates a `${logfile}.state` file, storing the position and creation time of the log file
+* Tries to aquire a file lock on the state file to ensure the file is only tailed once
+
+#### Delivery guarantees ack/nak
+* Only if the APM Server responds with a 200, the curret state is acknowledged an written to the `.state` file
+* If the APM Server returns an error code or if there's any network error, the previous state is restored (negative acknowledge)
+* If the APM Server can guaranee at-least-once delivery after returning a 200, the whole process has a at-least-once delivery guarantee
+
+#### Hanling of file rotation
+* The file might be rotated as we are reading it `apm.log` -> `apm-1.log` or `apm.log1`
+* As we keep the file handle open, renaming doesn't affect reading
+* If the agent restarts, it checks via the creation time which file to continue reading from
+  * If the file `${logfile}` has the same creation time we assume it has not been rotated
+  * If the creation time is different, search for a file in the same directory with that creation time
+
+
+### APM Server logs endpoint
+
+The endpoint would look similar to the current event intake API.
+
+Example:
+```bash
+curl -H 'Content-Type: application/x-ndjson' -XPOST localhost:8200/intake/v2/logs -d \
+'{"metadata":{"service":{"name":"my-service","process":{"pid":32146,"ppid":3372},"system":{"architecture":"x86_64","hostname":"localhost","platform":"Mac OS X"}}}
+{"metadata":{"file":{"path":"/var/log/apm.log","name":"apm.log"}}}
```

Comment:
This double metadata header stood out to me. Is the double metadata header what we want or something we're doing to avoid refactoring/changing too many things?

Reply:
This is intentional. It's not only a double metadata header. The 2nd type of metadata can appear multiple times in a single request and is valid until another metadata event comes along.

This is to support sending multiple log files in the same request. The metadata event allows providing file-specific metadata, like `file.name`, without having to change the events from the log file and without having to send repetitive information.